Add Downgrades to Monthly within the refund period

dzver opened the pull request:
#### Changes proposed in this Pull Request

This will allow users to downgrade to monthly as a solution to the problem that the plan was too expensive.

We currently offer Downgrades from Premium to Personal. This will offer downgrades from 1-year and 2-year Personal, Business, and eCommerce to the monthly versions of the same plan.

#### Testing instructions

1. Apply D71002-code
2. Buy a non-Premium annual or 2-year plan
3. Cancel it and select that the plan was too expensive in the cancellation survey
4. You should see an option to downgrade to monthly. Click that
5. The plan should be downgraded to monthly

Repeat the steps with a plan that has a domain. 
6. Make sure that the domain is kept no matter if the user selected to keep it or not.
7. Go to Store Admin, scroll down to the Payments section and make sure a partial refund with the expected amount happened

https://user-images.githubusercontent.com/82778/149167507-3c5a206b-f9ce-45c4-9fc2-d6822e1d2822.mov

EDIT: this isn't supposed to be visible after the refund period is over.
8. Now hardcode is_refundable on the backend to return false. Make sure the upsell is not visible

EDIT: 
9. After downgrading the plan to a monthly plan, repeat the process and cancel again, selecting that the plan was too expensive. There should be no errors.

dzver committed:
Add support for downgrades to monthly

This doesn't change the copy, it will be updated with a separate commit

dzver committed:
Update copy

In case users choose to downgrade, we will keep their domain. This has
to be reflected in the copy.

taggon commented:
This PR works as described. One thing I found, that may not be related to this, is "Manage Plan" button for Business monthly disappears after I switched the yearly plan to monthly through the survey form. Besides that, everything looks good to me.

<img width="1239" alt="Plans_‹_Taggon_on_wordpress_—_WordPress_com" src="https://user-images.githubusercontent.com/212034/149902422-c1210602-4b17-42a8-93cd-9d37609952c1.png">
<img width="1239" alt="Plans_‹_Taggon_on_wordpress_—_WordPress_com" src="https://user-images.githubusercontent.com/212034/149902461-fbee050b-cd16-42c0-9c6a-634f63eb5cba.png">

taggon reviewed (approved):
Please ship it if this PR doesn't cause the problem I mentioned above. We can fix it in a follow-up patch.

DavidRothstein commented on client/components/marketing-survey/cancel-purchase-form/step-components/downgrade-step.jsx:
```diff
@@ -68,6 +79,42 @@ export class DowngradeStep extends Component {
+				refundReason = (
+					<p>
+						{ translate(
+							'You will keep all features of your current plan including the domain %(domain)s but the plan period will be reduced.',
```
Probably it should say "most of the features" here (like below).  I think they still lose things like live chat?

DavidRothstein commented on client/components/marketing-survey/cancel-purchase-form/step-components/downgrade-step.jsx:
```diff
@@ -68,6 +79,42 @@ export class DowngradeStep extends Component {
+				refundReason = (
+					<p>
+						{ translate(
+							'You will keep most of the features of your current plan but will not have a free domain.'
```
Is "free domain" the terminology we typically use for this?  It's only free for the first year, and I thought we usually specify that.

DavidRothstein reviewed (commented):
I noticed one main bug: When downgrading a nonrefundable plan, it always says "You can downgrade and the new plan will renew at only $48.00" (which is equal to the **annual personal** plan cost, and which is the incorrect price).  I think it's coming from `getDowngradePlanRawPrice()`.

I also thought the downgrade wording was a little unexpected in the case where I have a domain:

1. If I choose "Cancel the plan, but keep [domain]" on the cancellation screen, then I see this during downgrade:

>You will keep most of the features of your current plan but will not have a free domain.

2. If I choose "Cancel the plan and the domain [domain]" on the cancellation screen, then I see this during downgrade:

>You will keep all features of your current plan including the domain [domain] but the plan period will be reduced.

I think it should show the second message in both cases, since in both cases I am keeping my existing domain.  (I guess it's more important to show that in the second case since in that scenario it is contradicting what I chose earlier, but it still seems like a more useful and less confusing message for the first case too.)

DavidRothstein commented:
>I noticed one main bug: When downgrading a nonrefundable plan...

Wait a minute - why is it even offering this as an option on nonrefundable plans in the first place?

If the plan is nonrefundable, then downgrading to monthly means you wind up with the same set of features you had before, and no money back... but suddenly you have an earlier expiration date.  That is not a choice that anyone would ever want to make :smile:  So unless we change the way that works, it probably shouldn't be an option on this screen.

See also previous discussion: pcbrnV-3Y8-p2#comment-5106

dzver committed:
Update copy

dzver committed:
Hide the upsell if the user can't refund

dzver commented:
I've addressed the issues regarding the canRefund with an if. This change isn't supposed to work after the refund period.

DavidRothstein commented:
It's still letting me downgrade a non-refundable plan to monthly.  This is what I see when downgrading a non-refundable Business plan (note the text at the top which is accidentally coming from the Premium-to-Personal downgrade path and which incorrectly implies I'll wind up with a Personal plan):

![downgrade-nonrefundable-business-plan](https://user-images.githubusercontent.com/235183/151449306-149a355f-4914-4e84-8bfe-ffd00d4b5b3c.png)

I think the code will need to hide the entire option in the case where the original plan is non-refundable.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7106684

Hi @dzver, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Switch to a monthly subscription`
- `Would you rather switch to the more affordable monthly subscription?`
- `You will keep most features of your current plan including the domain %(domain)s but the plan period will be reduced.`
- `You will keep most of the features of your current plan but will not have a free domain registration.`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7106684).

DavidRothstein commented:
That looks like the right idea, but it now crashes when trying to cancel a non-downgradable plan (for example, buy a monthly personal plan with a fake credit and try to cancel it; you can't fill out the cancellation survey due to `Uncaught TypeError: Cannot read properties of null (reading '0')`).

It seems like the `getRefundAmount()` helper doesn't work as expected for non-downgradable purchases.... so it would need to either be fixed to work better or its use avoided in that scenario.

dzver committed:
Address David's feedback on two more cases

There was a type error in selecting the plan was too expensive for
monthly plans;

The dialog was visible in some cases of a non-refundable plan

dzver commented:
@DavidRothstein thanks a lot for catching these cases. I updated the function with conditional chaining so that it doesn't break for refunding monthly plans.

DavidRothstein reviewed (approved):
Thanks -- seems to work as expected now.
